sklearn/exceptions.py

```diff
@@ -0,0 +1,135 @@
+    ...     print(repr(e))
+    ...                        # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+    NotFittedError('This LinearSVC instance is not fitted yet',)
+    """
+
+
+class FitFailedWarning(RuntimeWarning):
+    """Warning class used if there is an error while fitting the estimator.
+
+    This Warning is primarily used with meta estimators like GridSearchCV,
+    RandomizedSearchCV etc. to warn when there is an error while fitting the
+    estimator.
+
+    Examples
+    --------
+    >>> from sklearn.grid_search import GridSearchCV
+    >>> from sklearn.svm import LinearSVC
+    >>> from sklearn.exceptions import FitFailedWarning
+    >>> import warnings
+    >>> warnings.simplefilter('always', FitFailedWarning)
+    >>> gs = GridSearchCV(LinearSVC(), {'C': [-1, -2]}, error_score=0)
+    >>> X, y = [[1, 2], [3, 4], [5, 6], [7, 8], [8, 9]], [0, 0, 0, 1, 1]
+    >>> with warnings.catch_warnings(record=True) as w:
+    ...     try:
+    ...         gs.fit(X, y)   # This will raise a ValueError since C is < 0
+    ...     except ValueError:
+    ...         pass
+    ...     print("The warning raised was : %r" % w[-1].message)
+    ...                        # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+    The warning raised was : 'FitFailedWarning("Classifier fit failed. The
+    score on this train-test partition for these parameters will be set to
+    0.000000. Details: \\\\nValueError(\\'Penalty term must be positive; got
+    (C=-2)\\',)",)'
+    """
+
+
+class ChangedBehaviorWarning(UserWarning):
+    """Warning class used to notify the user of any change in the behavior.
+
+    Examples
+    --------
+    >>> from sklearn.grid_search import GridSearchCV
+    >>> from sklearn.svm import LinearSVC
+    >>> from sklearn.metrics import f1_score
+    >>> from sklearn.datasets import make_blobs
+    >>> from sklearn.exceptions import ChangedBehaviorWarning
+    >>> import warnings
+    >>> warnings.simplefilter('always', ChangedBehaviorWarning)
+    >>> gs = GridSearchCV(estimator=LinearSVC(random_state=0),
+    ...                   param_grid={'C': [1, 10, 100]}, scoring='f1_micro')
+    >>> X, y = make_blobs(random_state=0)
+    >>> with warnings.catch_warnings(record=True) as w:
+    ...     score = gs.fit(X[:20], y[:20]).score(X[20:], y[20:])
+    ...     print("The warning raised was : %r" % w[-1].message)
+    ...     print("The score is : %s" % score)
+    ...                        # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+    The warning raised was : 'ChangedBehaviorWarning("The long-standing
+    behavior to use the estimator\\'s score function in GridSearchCV.score
+    has changed. The scoring parameter is now used.",)'
+    The score is : 0.8625
+    """
+
+
+class DataConversionWarning(UserWarning):
+    """Warning used to notify implicit data conversions happening in the code.
+
+    This warning is raised when the data, usually the input, is implicitly
+    converted/copied based on the requirements.
```

Comment:
Let's clarify to ourself: are we warning the user because:
* had they chosen a better format the copy would have been unnecessary and hence memory efficiency is the issue?
* it is a `copy=False` operation so they've explicitly hinted that altering the existing array is okay, but we could not benefit from this optimisation?
* the copy is lossy in some way that might have been important to the user (e.g. large integers becoming low-precision floats)?
* they might have misunderstood the purpose of the function they are using, as evidenced by giving the wrong data type?
* some or all of the above